### PR TITLE
[CSA-58] Refactoring instant

### PR DESCRIPTION
**Pull request by OpenSuze**
**Please check if the PR fulfills these requirements** *(please use `'[x]'` to check the checkboxes, or submit the PR and then click the checkboxes)*
- [ ] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**Does this PR already have an issue describing the problem ?** *If so, link to this issue using `'#XXX'` and skip the rest*

**What kind of change does this PR introduce?** *(Bug fix, feature, docs update, ...)*

**What is the current behavior?** *(You can also link to an open issue here)*

**What is the new behavior (if this is a feature change)?**

**Does this PR introduce a breaking change?** *(What changes might users need to make in their application due to this PR?)*

**Other information**:

(if any of the questions/checkboxes don't apply, please delete them entirely)

### CODE REVIEW

File: data/crac-io/crac-io-json/src/test/java/com/farao_community/farao/data/crac_io_json/JsonSerializationConstantsTest.java [hunk 1]
```diff
@@ -125,13 +139,13 @@ private UsageRule mockUsageRule(Instant instant, UsageMethod usageMethod, String
...
+        UsageRule onInstant1 = mockUsageRule(INSTANT_PREV, UsageMethod.AVAILABLE, null, null, null, null, null);
+        UsageRule onInstant2 = mockUsageRule(INSTANT_PREV, UsageMethod.FORCED, null, null, null, null, null);
+        UsageRule onInstant3 = mockUsageRule(INSTANT_CURATIVE, UsageMethod.AVAILABLE, null, null, null, null, null);
 
-        assertEquals(comparator.compare(onInstant1, onInstant1), 0);
-        assertEquals(comparator.compare(onInstant2, onInstant2), 0);
-        assertEquals(comparator.compare(onInstant3, onInstant3), 0);
+        //assertEquals(0, comparator.compare(onInstant1, onInstant1));
+        //assertEquals(0, comparator.compare(onInstant2, onInstant2));
+        //assertEquals(0, comparator.compare(onInstant3, onInstant3));
```
Comment (pet-mit) [inline data/crac-io/crac-io-json/src/test/java/com/farao_community/farao/data/crac_io_json/JsonSerializationConstantsTest.java:148]: why did you comment this?

File: data/crac/crac-impl/src/test/java/com/farao_community/farao/data/crac_impl/InstantImplTest.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import com.farao_community.farao.data.crac_api.InstantKind;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * @author Joris Mancini {@literal <joris.mancini at rte-france.com>}
```
Comment (pet-mit) [inline data/crac/crac-impl/src/test/java/com/farao_community/farao/data/crac_impl/InstantImplTest.java:20]: no

File: data/crac/crac-impl/src/test/java/com/farao_community/farao/data/crac_impl/InstantImplTest.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright (c) 2019, RTE (http://www.rte-france.com)
```
Comment (pet-mit) [inline data/crac/crac-impl/src/test/java/com/farao_community/farao/data/crac_impl/InstantImplTest.java:2]:
```suggestion
 * Copyright (c) 2023, RTE (http://www.rte-france.com)
```

File: data/crac/crac-impl/src/test/resources/TestCase2Nodes.xiidm
Comment (pet-mit) [inline data/crac/crac-impl/src/test/resources/TestCase2Nodes.xiidm:1]: why has this changed ? please revert unnecessary chnages, they make reviewing changes very difficult (indentation, iidm, line breaks, changing location of methods in class file...)

File: data/crac/crac-util/src/main/java/com/farao_community/farao/data/crac_util/CracValidator.java [hunk 4]
```diff
@@ -47,25 +44,28 @@ public static List<String> validateCrac(Crac crac, Network network) {
...
+                        duplicateCnecOnOutageInstant(crac, cnec);
+                        report.add(String.format("CNEC \"%s\" has no associated automaton. It will be cloned on the OUTAGE instant in order to be secured during preventive RAO.", cnec.getId()));
+                    });
+            });
+        }
         return report;
     }
 
     private static void duplicateCnecOnOutageInstant(Crac crac, FlowCnec cnec) {
+        Instant outageInstant = getOutageInstant(cnec);
```
Comment (pet-mit) [inline data/crac/crac-util/src/main/java/com/farao_community/farao/data/crac_util/CracValidator.java:68]: crac.getInstant(OUTAGE)

File: data/rao-result/rao-result-json/src/main/java/com/farao_community/farao/data/rao_result_json/RaoResultJsonConstants.java [hunk 5]
```diff
@@ -197,57 +197,39 @@ public static Side deserializeSide(String stringValue) {
...
-                return PREVENTIVE_INSTANT;
-            case OUTAGE:
-                return OUTAGE_INSTANT;
-            case AUTO:
-                return AUTO_INSTANT;
-            case CURATIVE:
-                return CURATIVE_INSTANT;
-            default:
-                throw new FaraoException(String.format("Unsupported instant %s", instant));
+            return "";
```
Comment (pet-mit) [inline data/rao-result/rao-result-json/src/main/java/com/farao_community/farao/data/rao_result_json/RaoResultJsonConstants.java:202]:
```suggestion
            return INITIAL_INSTANT;
```

File: data/rao-result/rao-result-json/src/main/java/com/farao_community/farao/data/rao_result_json/RaoResultJsonConstants.java [hunk 6]
```diff
@@ -197,57 +197,39 @@ public static Side deserializeSide(String stringValue) {
...
-                return Instant.OUTAGE;
-            case AUTO_INSTANT:
-                return Instant.AUTO;
-            case CURATIVE_INSTANT:
-                return Instant.CURATIVE;
-            default:
-                throw new FaraoException(String.format("Unrecognized instant %s", stringValue));
+    public static String deserializeInstantId(String stringValue) {
+        // TODO review this
+        if (Objects.equals(stringValue, "") || Objects.equals(stringValue, INITIAL_INSTANT)) {
```
Comment (pet-mit) [inline data/rao-result/rao-result-json/src/main/java/com/farao_community/farao/data/rao_result_json/RaoResultJsonConstants.java:210]:
```suggestion
        if (stringValue.equals(INITIAL_INSTANT)) {
```

File: data/rao-result/rao-result-json/src/main/java/com/farao_community/farao/data/rao_result_json/serializers/AngleCnecResultArraySerializer.java [hunk 7]
```diff
@@ -33,45 +34,48 @@ static void serialize(RaoResult raoResult, Crac crac, JsonGenerator jsonGenerato
...
     }
 
-    private static void serializeAngleCnecResultForOptimizationState(Instant optInstant, AngleCnec angleCnec, RaoResult raoResult, JsonGenerator jsonGenerator) throws IOException {
+    private static void serializeAngleCnecResultForOptimizationState(String optInstantId, AngleCnec angleCnec, RaoResult raoResult, JsonGenerator jsonGenerator) throws IOException {
 
-        if (containsAnyResultForOptimizationState(raoResult, angleCnec, optInstant)) {
-            jsonGenerator.writeObjectFieldStart(serializeInstant(optInstant));
-            serializeAngleCnecResultForOptimizationStateAndUnit(optInstant, Unit.DEGREE, angleCnec, raoResult, jsonGenerator);
+        if (containsAnyResultForOptimizationState(raoResult, angleCnec, optInstantId)) {
+            jsonGenerator.writeObjectFieldStart(Optional.ofNullable(optInstantId).orElse("")); // TODO use serializer ?
```
Comment (pet-mit) [inline data/rao-result/rao-result-json/src/main/java/com/farao_community/farao/data/rao_result_json/serializers/AngleCnecResultArraySerializer.java:69]: yes, use serializer

**Review by pet-mit — CHANGES_REQUESTED** [review]
please revert unnecessary changes to make reviewing less difficult

File: data/crac-creation/crac-creation-util/src/main/java/com/farao_community/farao/data/crac_creation/util/ucte/UcteFlowElementHelper.java [hunk 8]
```diff
@@ -181,7 +180,7 @@ private void interpretDanglingLine(DanglingLine danglingLine) {
             this.connectableIdInNetwork = tieLine.getId();
             Branch.Side side = tieLine.getDanglingLine1() == danglingLine ? Branch.Side.ONE : Branch.Side.TWO;
             // dangling line convention is x node to terminal, so dl 1 is towards terminal 1 (opposite) and dl 2 is towards terminal 2 (direct)
-            this.isInvertedInNetwork = tieLine.getDanglingLine1() == danglingLine ? !isInvertedInNetwork : isInvertedInNetwork;
+            this.isInvertedInNetwork = (tieLine.getDanglingLine1() == danglingLine) != isInvertedInNetwork;
```
[the remark this answers is not in the document]
Comment (OpenSuze) [reply data/crac-creation/crac-creation-util/src/main/java/com/farao_community/farao/data/crac_creation/util/ucte/UcteFlowElementHelper.java:183]: fixed

File: data/crac-io/crac-io-json/src/test/java/com/farao_community/farao/data/crac_io_json/JsonSerializationConstantsTest.java (hunk 1 above)
Comment (OpenSuze) [reply data/crac-io/crac-io-json/src/test/java/com/farao_community/farao/data/crac_io_json/JsonSerializationConstantsTest.java:148]: I forgot to remove those comments

File: data/crac/crac-impl/src/test/resources/TestCase2Nodes.xiidm
Comment (OpenSuze) [reply data/crac/crac-impl/src/test/resources/TestCase2Nodes.xiidm:1]: Reverted unnecessary changes

**Review by pet-mit — CHANGES_REQUESTED** [review]
Some things are still to change, but overall very good work, it was a tough change ! :)

**Comment by OpenSuze** [comment]
Tous les tests cucumber passent !

File: data/crac-io/crac-io-json/src/main/java/com/farao_community/farao/data/crac_io_json/serializers/InstantSerializer.java [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+ */
+public class InstantSerializer extends AbstractJsonSerializer<Instant> {
+
+    @Override
+    public void serialize(Instant value, JsonGenerator gen, SerializerProvider serializers) throws IOException {
+        gen.writeStartObject();
+
+        gen.writeStringField(ID, value.getId());
+
+        gen.writeStringField(INSTANT_KIND, value.getKind().toString());
```
Comment (pet-mit) [inline data/crac-io/crac-io-json/src/main/java/com/farao_community/farao/data/crac_io_json/serializers/InstantSerializer.java:30]: don't use toString

File: data/crac-io/crac-io-json/src/test/java/com/farao_community/farao/data/crac_io_json/JsonRetrocompatibilityTest.java [hunk 10]
```diff
@@ -249,10 +244,30 @@ void importV1Point9Test() {
         testContentOfV1Point9Crac(crac);
     }
 
+    @Test
+    void importV1Point10Test() {
+        // Add support for user-defined Instants
+        InputStream cracFile = getClass().getResourceAsStream("/retrocompatibility/v1/crac-v1.10.json");
+
+        Crac crac = new JsonImport().importCrac(cracFile);
```
Comment (pet-mit) [inline data/crac-io/crac-io-json/src/test/java/com/farao_community/farao/data/crac_io_json/JsonRetrocompatibilityTest.java:252]: you should test the instants of the CRAC, which is the most important feature of this new version

File: data/crac/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/CracImpl.java [hunk 11]
```diff
@@ -187,51 +186,61 @@ public Instant getInstant(InstantKind instantKind) {
...
-    public Instant getInstantBefore(String providedInstantId) {
-        if (!instants.containsKey(providedInstantId)) {
-            throw new FaraoException(String.format("Instant '%s' has not been defined", providedInstantId));
+    public Instant getOutageInstant() {
+        return getInstant(InstantKind.OUTAGE);
+    }
+
+    @Override
+    public boolean hasAutoInstant() {
+        return getInstants(InstantKind.AUTO).size() == 1;
```
Comment (pet-mit) [inline data/crac/crac-impl/src/main/java/com/farao_community/farao/data/crac_impl/CracImpl.java:227]: make it generic
```suggestion
        return getInstants(InstantKind.AUTO).size() >= 1;
```

File: data/rao-result/rao-result-impl/src/main/java/com/farao_community/farao/data/rao_result_impl/RaoResultImpl.java [hunk 12]
```diff
@@ -312,7 +312,8 @@ private State stateBefore(State state) {
     }
 
     private State stateBefore(String contingencyId, Instant instant) {
-        if (instant.comesBefore(crac.getInstant(InstantKind.AUTO))) {
+        Instant otherInstant = crac.hasAutoInstant() ? crac.getInstant(InstantKind.AUTO) : crac.getInstant(InstantKind.CURATIVE);
+        if (instant.comesBefore(otherInstant)) {
```
Comment (pet-mit) [inline data/rao-result/rao-result-impl/src/main/java/com/farao_community/farao/data/rao_result_impl/RaoResultImpl.java:316]: all this is strictly equivalent to checking if instant is OUTAGE

File: data/rao-result/rao-result-json/src/main/java/com/farao_community/farao/data/rao_result_json/RaoResultJsonConstants.java [hunk 13]
```diff
@@ -201,37 +203,46 @@ public static String serializeInstantId(Instant instant) {
...
             return INITIAL_INSTANT_ID;
         }
-        // TODO review this
         return instant.getId();
     }
 
     public static String deserializeInstantId(String stringValue) {
-        if (stringValue.equals("") || stringValue.equals(INITIAL_INSTANT_ID)) { // TODO review this "" test (see RaoResultRoundTripTest)
+        return stringValue;
+    }
```
Comment (pet-mit) [inline data/rao-result/rao-result-json/src/main/java/com/farao_community/farao/data/rao_result_json/RaoResultJsonConstants.java:211]: we don't need this anymore

File: data/rao-result/rao-result-json/src/main/java/com/farao_community/farao/data/rao_result_json/RaoResultJsonConstants.java [hunk 14]
```diff
@@ -201,37 +203,46 @@ public static String serializeInstantId(Instant instant) {
...
+        if (primaryVersionNumber <= 1 && subVersionNumber <= 3) {
             switch (stringValue) {
                 case INITIAL_OPT_STATE:
-                    return null;
+                    return INITIAL_INSTANT_ID;
                 case AFTER_PRA_OPT_STATE:
                     return PREVENTIVE_INSTANT_ID;
                 case AFTER_ARA_OPT_STATE:
-                    return AUTO_INSTANT_ID;
+                    return (primaryVersionNumber == 1 && subVersionNumber == 1 && cracDoesNotContainAutoRemedialAction(crac)) ? PREVENTIVE_INSTANT_ID : AUTO_INSTANT_ID;
```
Comment (pet-mit) [inline data/rao-result/rao-result-json/src/main/java/com/farao_community/farao/data/rao_result_json/RaoResultJsonConstants.java:231]: can this work ?
```suggestion
                    return (primaryVersionNumber == 1 && subVersionNumber == 1 && !crac.hasAutoInstant()) ? PREVENTIVE_INSTANT_ID : AUTO_INSTANT_ID;
```

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/search_tree/inputs/SearchTreeInput.java [hunk 15]
```diff
@@ -124,14 +133,20 @@ public SearchTreeInputBuilder withToolProvider(ToolProvider toolProvider) {
             return this;
         }
 
+        public SearchTreeInputBuilder withCrac(Crac crac) {
+            this.crac = crac;
+            return this;
+        }
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/search_tree/inputs/SearchTreeInput.java:139]: you only need the outage instant, so replace this with an "withOutageInstant" method

File: ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/linear_optimisation/inputs/IteratingLinearOptimizerInput.java [hunk 16]
```diff
@@ -180,6 +189,11 @@ public IteratingLinearOptimizerInputBuilder withToolProvider(ToolProvider toolPr
             return this;
         }
 
+        public IteratingLinearOptimizerInputBuilder withCrac(Crac crac) {
+            this.crac = crac;
+            return this;
+        }
```
Comment (pet-mit) [inline ra-optimisation/search-tree-rao/src/main/java/com/farao_community/farao/search_tree_rao/linear_optimisation/inputs/IteratingLinearOptimizerInput.java:195]: same

**Review by pet-mit — CHANGES_REQUESTED** [review]

File: data/crac-io/crac-io-json/src/test/java/com/farao_community/farao/data/crac_io_json/JsonRetrocompatibilityTest.java (hunk 10 above)
Comment (OpenSuze) [reply data/crac-io/crac-io-json/src/test/java/com/farao_community/farao/data/crac_io_json/JsonRetrocompatibilityTest.java:252]: I tested this in the end of the file, is it ok ?

File: data/crac-io/crac-io-json/src/main/java/com/farao_community/farao/data/crac_io_json/deserializers/CracDeserializer.java [hunk 17]
```diff
@@ -162,7 +162,8 @@ private void scrollJsonUntilField(JsonParser jsonParser, String field) throws IO
     private void checkVersion(String cracVersion) {
 
         if (getPrimaryVersionNumber(CRAC_IO_VERSION) > getPrimaryVersionNumber(cracVersion)) {
-            throw new FaraoException(String.format("CRAC importer %s is no longer compatible with json CRAC version %s", CRAC_IO_VERSION, cracVersion));
+            LOGGER.warn("CRAC importer {} might not be longer compatible with json CRAC version {}", CRAC_IO_VERSION, cracVersion);
```
Comment (pet-mit) [inline data/crac-io/crac-io-json/src/main/java/com/farao_community/farao/data/crac_io_json/deserializers/CracDeserializer.java:165]: ok for me. It means that we can drop support for versions 1 from now on

File: data/crac-io/crac-io-json/src/test/java/com/farao_community/farao/data/crac_io_json/JsonRetrocompatibilityTest.java (hunk 10 above)
Comment (pet-mit) [reply data/crac-io/crac-io-json/src/test/java/com/farao_community/farao/data/crac_io_json/JsonRetrocompatibilityTest.java:252]: ok sorry i forgot you put it there

**Review by pet-mit — CHANGES_REQUESTED** [review]
just remove the TODO

**Review by pet-mit — APPROVED** [review]

**Pull request merged**